### DISCUSSION TITLE
TZ3000 4 button switch - Double and hold not supported. Help needed to implement.

### DISCUSSION THREAD

**Discussion by deeja** [Koenkk/zigbee2mqtt#8852] [Devices]
Hey I bought a generic 4 button switch which you can see in operation here: https://www.youtube.com/watch?v=Hf86De6qWyA
This is the fully working SMARTTHINGS code that goes with it (so I'm told)
https://github.com/YooSangBeom/SangBoyST/blob/master/devicetypes/sangboy/zemismart-button.src/zemismart-button.groovy

The current versions of z2m support only the push operation (single push). I'd like to add in the code to support the other two modes, but really have no idea where to start.

Any help on where to start would be appreciated.

![image](https://user-images.githubusercontent.com/174247/134440749-010e3adf-a3b0-4fbf-b736-15764bb4a625.png)

**Comment by deeja** [comment]
I'll also mention that the particular switch I have is a `rsh-zigbee-sc04` that has the following debug info when pressed:
```
Zigbee2MQTT:debug 2021-09-23 02:55:23: Received Zigbee message from '0x5c0272fffe24b6f8', type 'commandStep', cluster 'genLevelCtrl', data '{"stepmode":1,"stepsize":51,"transtime":10}' from endpoint 1 with groupID 0
Zigbee2MQTT:info  2021-09-23 02:55:23: MQTT publish: topic 'zigbee2mqtt/0x5c0272fffe24b6f8', payload '{"action":"brightness_step_down","action_step_size":51,"action_transition_time":0.1,"battery":100,"linkquality":132}'
Zigbee2MQTT:info  2021-09-23 02:55:23: MQTT publish: topic 'zigbee2mqtt/0x5c0272fffe24b6f8', payload '{"action":"","battery":100,"linkquality":132}'
Zigbee2MQTT:info  2021-09-23 02:55:23: MQTT publish: topic 'zigbee2mqtt/0x5c0272fffe24b6f8/action', payload 'brightness_step_down'
```

**Comment by deeja** [comment]
```json
{
    "date_code": "",
    "definition": {
        "description": "Wireless switch with 4 buttons",
        "exposes": [{
            "access": 1,
            "description": "Remaining battery in %",
            "name": "battery",
            "property": "battery",
            "type": "numeric",
            "unit": "%",
            "value_max": 100,
            "value_min": 0
        }, {
            "access": 1,
            "description": "Triggered action (e.g. a button click)",
            "name": "action",
            "property": "action",
            "type": "enum",
            "values": ["on", "off", "brightness_move_up", "brightness_step_up", "brightness_step_down", "brightness_move_down", "brightness_stop"]
        }, {
            "access": 1,
            "description": "Link quality (signal strength)",
            "name": "linkquality",
            "property": "linkquality",
            "type": "numeric",
            "unit": "lqi",
            "value_max": 255,
            "value_min": 0
        }],
        "model": "TS004F",
        "supports_ota": false,
        "vendor": "TuYa"
    },
    "endpoints": {
        "1": {
            "bindings": [{
                "cluster": "genPowerCfg",
                "target": {
                    "endpoint": 1,
                    "ieee_address": "0xccccccfffea5f198",
                    "type": "endpoint"
                }
            }],
            "clusters": {
                "input": ["genBasic", "genPowerCfg", "genIdentify", "genGroups", "genOnOff", "touchlink"],
                "output": ["genOta", "genTime", "genIdentify", "genGroups", "genScenes", "genOnOff", "genLevelCtrl", "touchlink"]
            },
            "configured_reportings": [{
                "attribute": "batteryPercentageRemaining",
                "cluster": "genPowerCfg",
                "maximum_report_interval": 62000,
                "minimum_report_interval": 3600,
                "reportable_change": 0
            }]
        }
    },
    "friendly_name": "0x5c0272fffe24b6f8",
    "ieee_address": "0x5c0272fffe24b6f8",
    "interview_completed": true,
    "interviewing": false,
    "manufacturer": "_TZ3000_xabckq1v",
    "model_id": "TS004F",
    "network_address": 5967,
    "power_source": "Battery",
    "supported": true,
    "type": "EndDevice"
}
```

**Comment by deeja** [comment]
There is some Tasmota related info here https://giters.com/arendst/Tasmota/issues/13128#issuecomment-919442480

**Comment by ghost** [comment]
There's already a [very long and active discussion about this](https://github.com/Koenkk/zigbee2mqtt/discussions/7158). Feel free to tinker with the work already done and see if you can add to it and figure it out, I'm stuck on it too.

Have you tested it with smartthings? I had figured from [discussions on their own forums](https://community.smartthings.com/t/four-button-tuya-zigbee-device-dth-dth-link-in-post-7-same-device-from-multiple-brands-zemismart-yagusmart-losonho-dyhf-etc/206149/10) that they also had the same problems, but if it's the case that they have it working that would be good news.

**Reply by deeja** [reply]
Thanks for that. Looks like it's more complicated than I expected, but also it does look like Smartthings is working from the video (although I haven't tried it myself). 
Cheers for the redirect.

**Comment by kkossev** [comment]
In SmartThings it works the same way as in Z2M - single press for all 4 buttons, hold/release only for the right two buttons. No double press events are generated in the default operational mode, that TS004F is initialized after pairing to any Zigbee coordinator different than Tuya.

**Reply by deeja** [reply]
Are you sure? Looks like he's using double press in this video https://youtu.be/Hf86De6qWyA?t=85 
And the code here references a double + hold https://github.com/YooSangBeom/SangBoyST/blob/master/devicetypes/sangboy/zemismart-button.src/zemismart-button.groovy#L61

**Reply by deeja** [reply]
Ah, yeah it's referenced here by the looks. Push only 
https://github.com/YooSangBeom/SangBoyST/blob/master/devicetypes/sangboy/zemismart-button.src/zemismart-button.groovy#L456
```
  if(device.getDataValue("model") == "TS004F")
    {
        createChildButtonDevices_1(numberOfButtons) //
        sendEvent(name: "supportedButtonValues", value: ["pushed"].encodeAsJSON(), displayed: false)
    }
    else
    {
        createChildButtonDevices(numberOfButtons) //Todo
        sendEvent(name: "supportedButtonValues", value: ["pushed","held","double"].encodeAsJSON(), displayed: false)
    }
```

**Comment by kkossev** [comment]
These are the two different operational modes, that can be configured only via Tuya gateway (for now ) :
![image](https://user-images.githubusercontent.com/6189950/134812615-2f3563d1-1a9a-4b42-9c96-4f578667cbd9.png)

**Reply by kkossev** [reply]
... and what was found recently is that the mode change works (or could work) in other platforms also, **but** - only if the TS004F switch was initially paired to Tuya hub! :( 
If the switch was never paired to Tuya hub or the batteries were removed for several minutes (results in some kind of 'cold boot'?) - then the switch mode commands are acknowledged with 'success' result code, but nothing happens (mode is not actually changed).
